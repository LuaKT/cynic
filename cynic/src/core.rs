#![allow(dead_code, unused_variables, missing_docs)]

// TODO: Everything in here is actually typed.  Need an untyped core with this
// layered on top...

use std::marker::PhantomData;

// Annoyingly this means people can't derive Deserialize _as well as_ use cynics derives.
// But whatever, don't do that people?  I _think_ it's an OK limitation.
pub trait QueryFragment<'de>: serde::Deserialize<'de> {
    type SchemaType;

    fn query(builder: QueryBuilder<Self::SchemaType>);
}

impl<'de, T> QueryFragment<'de> for Option<T>
where
    T: QueryFragment<'de>,
{
    type SchemaType = Option<T::SchemaType>;

    fn query(builder: QueryBuilder<Self::SchemaType>) {
        T::query(builder.into_inner())
    }
}

impl<'de, T> QueryFragment<'de> for Vec<T>
where
    T: QueryFragment<'de>,
{
    type SchemaType = Vec<T::SchemaType>;

    fn query(builder: QueryBuilder<Self::SchemaType>) {
        T::query(builder.into_inner())
    }
}

impl<'de> QueryFragment<'de> for bool {
    type SchemaType = bool;

    fn query(builder: QueryBuilder<Self::SchemaType>) {}
}

// TODO: Can I also impl this for &'static str?
impl<'de> QueryFragment<'de> for String {
    type SchemaType = String;

    fn query(builder: QueryBuilder<Self::SchemaType>) {}
}

// TODO: impl QueryFragment for Option, Box etc.

// TODO: QueryBuilder or SelectionBuilder?
pub struct QueryBuilder<'a, SchemaType> {
    phantom: PhantomData<fn() -> SchemaType>,
    fields: &'a mut Vec<Field>,
}

impl<'a, T> QueryBuilder<'a, Vec<T>> {
    fn into_inner(self) -> QueryBuilder<'a, T> {
        QueryBuilder {
            fields: self.fields,
            phantom: PhantomData,
        }
    }
}

impl<'a, T> QueryBuilder<'a, Option<T>> {
    fn into_inner(self) -> QueryBuilder<'a, T> {
        QueryBuilder {
            fields: self.fields,
            phantom: PhantomData,
        }
    }
}

#[derive(Debug)]
pub struct Field {
    name: &'static str,
    children: Vec<Field>,
}

impl<'a, SchemaType> QueryBuilder<'a, SchemaType> {
    // TODO: this is just for testing
    pub fn temp_new(fields: &'a mut Vec<Field>) -> Self {
        QueryBuilder {
            phantom: PhantomData,
            fields,
        }
    }

    pub fn select_field<'b, FieldMarker, FieldType>(
        &'b mut self,
    ) -> FieldSelectionBuilder<'b, FieldType>
    where
        FieldMarker: FieldName,
        SchemaType: HasField<FieldMarker, FieldType>,
    {
        self.fields.push(Field {
            name: FieldMarker::name(),
            children: Vec::new(),
        });

        FieldSelectionBuilder {
            field: self.fields.last_mut().unwrap(),
            phantom: PhantomData,
        }
    }

    // TODO: FragmentSpread & InlineFragment go here...

    // TODO: Could done be done via drop?  Maybe.
    pub fn done(self) {}
}

pub struct FieldSelectionBuilder<'a, SchemaType> {
    phantom: PhantomData<fn() -> SchemaType>,
    field: &'a mut Field,
}

impl<'a, FieldSchemaType> FieldSelectionBuilder<'a, FieldSchemaType> {
    // Note: I'm assuming that the DSL will use $whatever for variables
    // which provides a name and a way to do magic on that name to make
    // it work in all the contexts I need.
    //
    // Although the way I've ended up writing this - I may actually
    // keep the existing behaviour for arguments where each arg ends up in
    // a variable in the eventual query even if they _could_ be coded into the
    // document. Think it's easier, and not aware of any downsides.  Although
    // could possibly look into $whatever magic later...
    //
    // Though that's technically not possible is it.  Because we can't
    // serialize values at this point without knowing the destination format,
    // unless we do some serde transcoding type thing.
    // Maybe that's the only option?

    // TODO: Need to work on typesafety here...
    //pub fn argument(&mut self, name: &'static str, value: InputValue) {}

    // TODO: probably need all the reference nonsense here for value type
    pub fn argument<ArgumentName, ValueType>(&mut self, value: ValueType)
    where
        FieldSchemaType: HasArgument<ArgumentName>,
        ValueType: InputValue<FieldSchemaType::ArgumentSchemaType>,
    {
        todo!()
    }

    // Note: this is old code - actually I don't want field unwrapping because I care about
    // the options & vecs
    //
    // pub fn select_children<'b>(&'b mut self) -> QueryBuilder<'b, FieldSchemaType::InnerNamedType>
    // where
    //     FieldSchemaType: CompositeFieldType,

    pub fn select_children<'b>(&'b mut self) -> QueryBuilder<'b, FieldSchemaType> {
        QueryBuilder {
            phantom: PhantomData,
            fields: &mut self.field.children,
        }
    }

    // TODO: probably need an alias function here that defines an alias.

    // TODO: Could done be done via drop?  Maybe.
    pub fn done(self) {}
}

pub trait FieldName {
    fn name() -> &'static str;
}

pub trait HasField<FieldMarker, FieldType> {}

pub trait HasArgument<ArgumentName> {
    type ArgumentSchemaType;

    fn name() -> &'static str;
}

// TODO: Name of this vs the actual Value type I want to output
pub trait InputValue<SchemaType> {
    // TODO: Bet the self type & references are going to be a PITA with this...
    // fn to_actual_value(&self) -> ();
}

impl InputValue<crate::Id> for crate::Id {}
impl InputValue<Option<crate::Id>> for crate::Id {}
impl InputValue<Option<crate::Id>> for Option<crate::Id> {}
impl InputValue<Vec<crate::Id>> for crate::Id {}
impl InputValue<Vec<crate::Id>> for Vec<crate::Id> {}
// TODO: etc.

// TODO: CompositeFieldType vs CompositeType?
pub trait CompositeFieldType {
    // TODO: DO I even need this...
    type InnerNamedType;
}

impl<T> CompositeFieldType for Box<T>
where
    T: CompositeFieldType,
{
    type InnerNamedType = T;
}

impl<T> CompositeFieldType for Vec<T>
where
    T: CompositeFieldType,
{
    type InnerNamedType = T;
}

impl<T> CompositeFieldType for Option<T>
where
    T: CompositeFieldType,
{
    type InnerNamedType = T;
}
