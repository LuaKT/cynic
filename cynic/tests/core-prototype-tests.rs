#![allow(dead_code)]

use cynic::core::QueryFragment;
use cynic_proc_macros::QueryFragment2;
use serde::Deserialize;

mod schema {
    cynic::use_schema_2!("tests/test-schema.graphql");
}

mod manual_schema {
    use cynic::core;

    pub struct Query {}

    pub mod query_fields {
        use cynic::core;

        pub struct AllPosts {}

        pub struct Post {}

        impl core::FieldName for AllPosts {
            fn name() -> &'static str {
                "allPosts"
            }
        }

        impl core::FieldName for Post {
            fn name() -> &'static str {
                "post"
            }
        }

        impl core::HasField<AllPosts, Vec<super::BlogPost>> for super::Query {}
        impl core::HasField<Post, Option<super::BlogPost>> for super::Query {}

        pub mod post_arguments {
            use cynic::core;

            pub struct Id {}

            impl core::HasArgument<Id> for super::Post {
                type ArgumentSchemaType = cynic::Id;

                fn name() -> &'static str {
                    "id"
                }
            }
        }
    }

    pub struct BlogPost;
    pub mod blog_post_fields {
        use cynic::core;

        pub struct Author;

        pub struct HasMetadata;

        impl core::FieldName for Author {
            fn name() -> &'static str {
                "author"
            }
        }

        impl core::FieldName for HasMetadata {
            fn name() -> &'static str {
                "hasMetadata"
            }
        }

        // Note: The schema module could also (probably) output HasField impls for
        // the various option types that are also valid for the given type?
        impl core::HasField<Author, super::Author> for super::BlogPost {}

        impl core::HasField<HasMetadata, Option<bool>> for super::BlogPost {}
    }

    // Note: this might not be needed..
    impl core::CompositeFieldType for BlogPost {
        type InnerNamedType = BlogPost;
    }

    pub struct Author;
    pub mod author_fields {
        use cynic::core;

        pub struct Name;

        impl core::FieldName for Name {
            fn name() -> &'static str {
                "name"
            }
        }

        impl core::HasField<Name, Option<String>> for super::Author {}
    }

    // Note: this might not be needed..
    impl core::CompositeFieldType for Author {
        type InnerNamedType = Author;
    }
}

#[derive(cynic::QueryFragment2, Debug)]
#[cynic(
    schema_path = "tests/test-schema.graphql",
    schema_module = "schema",
    graphql_type = "Query"
)]
// #[derive(serde::Deserialize)]
struct MyQuery {
    #[arguments(id = "TODO")]
    post: Option<BlogPostOutput>,
    #[cynic(rename = "allPosts")]
    posts: Vec<BlogPostOutput>,
}

// impl<'de> cynic::core::QueryFragment<'de> for MyQuery {
//     type SchemaType = schema::Query;

//     fn query(mut builder: cynic::core::QueryBuilder<Self::SchemaType>) {
//         let mut post_field_builder =
//             builder.select_field::<schema::query_fields::Post, <Option<BlogPostOutput> as QueryFragment>::SchemaType>();

//         let post_builder = post_field_builder.select_children();

//         <Option<BlogPostOutput> as QueryFragment>::query(post_builder);

//         post_field_builder.done();

//         let mut post_list_field_builder =
//             builder.select_field::<schema::query_fields::AllPosts, <Vec<BlogPostOutput> as QueryFragment>::SchemaType>();

//         let post_list_builder = post_list_field_builder.select_children();

//         <Vec<BlogPostOutput> as QueryFragment>::query(post_list_builder);

//         post_list_field_builder.done()
//     }
// }

#[derive(cynic::QueryFragment2, Debug)]
#[cynic(
    schema_path = "tests/test-schema.graphql",
    schema_module = "schema",
    graphql_type = "BlogPost"
)]
struct BlogPostOutput {
    has_metadata: Option<bool>,
    author: AuthorOutput,
}

// impl<'de> cynic::core::QueryFragment<'de> for BlogPostOutput {
//     type SchemaType = schema::BlogPost;

//     fn query(mut builder: cynic::core::QueryBuilder<Self::SchemaType>) {
//         builder
//             .select_field::<schema::blog_post_fields::HasMetadata, <Option<bool> as QueryFragment>::SchemaType>()
//             .done();

//         let mut author_field_builder = builder
//             .select_field::<schema::blog_post_fields::Author, <AuthorOutput as QueryFragment>::SchemaType>();

//         let author_builder = author_field_builder.select_children();

//         AuthorOutput::query(author_builder);

//         author_field_builder.done();
//         builder.done();
//     }
// }

#[derive(cynic::QueryFragment2, Debug)]
#[cynic(
    schema_path = "tests/test-schema.graphql",
    schema_module = "schema",
    graphql_type = "Author"
)]
struct AuthorOutput {
    name: Option<String>,
}

// impl<'de> cynic::core::QueryFragment<'de> for AuthorOutput {
//     type SchemaType = schema::Author;

//     fn query(mut builder: cynic::core::QueryBuilder<Self::SchemaType>) {
//         builder
//             .select_field::<schema::author_fields::Name, Option<String>>()
//             .done();

//         builder.done();
//     }
// }

#[test]
fn test_using_this_shit() {
    use cynic::core::{QueryBuilder, QueryFragment};

    let mut fields = Vec::new();

    let builder: QueryBuilder<schema::Query> = QueryBuilder::temp_new(&mut fields);

    MyQuery::query(builder);

    insta::assert_debug_snapshot!(fields);
}

#[test]
fn test_deserialize() {
    use serde_json::json;

    let data = json!({
        "post": {"hasMetadata": false, "author": {"name": "Not Me"}},
        "allPosts": [{"hasMetadata": true, "author": {"name": "Me"}}]
    });

    insta::assert_debug_snapshot!(serde_json::from_value::<MyQuery>(data).unwrap());
}
