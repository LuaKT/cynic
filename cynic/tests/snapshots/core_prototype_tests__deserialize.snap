---
source: cynic/tests/core-prototype-tests.rs
expression: "serde_json::from_value::<MyQuery>(data).unwrap()"

---
MyQuery {
    post: Some(
        BlogPostOutput {
            has_metadata: Some(
                false,
            ),
            author: AuthorOutput {
                name: Some(
                    "Not Me",
                ),
            },
        },
    ),
    posts: [
        BlogPostOutput {
            has_metadata: Some(
                true,
            ),
            author: AuthorOutput {
                name: Some(
                    "Me",
                ),
            },
        },
    ],
}
